### ISSUE TITLE
Get client disconnect notifications in bidirectional async gRPC

### ISSUE THREAD

**Issue by tobias-** [line/armeria#971]
We don't get any notification when the client closes the connection when using asynchronous bidirectional stream calls.

https://github.com/SourceForgery/tachikoma/blob/ddfb03ba5d25f3de61e8fe1e1b2cb9be2aba0eb4/tachikoma-grpc/src/main/kotlin/com/sourceforgery/tachikoma/mta/MTAEmailQueueService.kt#L31

We get neither `onComplete()`, nor `onError()` in the ~ResponseObserver~`StreamObserver` when the client closes the socket.

Just to clarify the code, the `getEmails()` method will return immediately, but will hook up a listener to the RabbitMQ which will then deliver messages via `responseObserver.onNext(response)`. The RabbitMQ will stop delivering messages when the future is finished (which will only happen in response to exceptions or close).

The lifetime of this call is counted in months, as it's more of a subscription than anything else.

**Comment by anuraaga** [comment]
It's a bit of a weird syntax, but I think ` ((ServerCallStreamObserver<Response>) observer).setOnCancelHandler(() -> System.out.println("cancelled."));` is supposed to be called on client close can you check if you see that?

**Comment by tobias-** [comment]
Unfortunately it doesn't seem to work. The `ArmeriaServerCall.closeListener()` is never called, and neither is anything else in the `ArmeriaServerCall` class, as far as I can tell.

It seems as though code is missing from either the `Http2ObjectEncoder.doClose()` or the `HttpServerHandler.channelInactive(...)` as those are the only places that Armeria gets notified that the connection is closed?

**Comment by anuraaga** [comment]
Thanks for checking will see if I can get a simple repro. A team member of mine looked at this before and I think had good luck with the cancel handler but I think that was a streaming response with no request streaming.

I'd expect this to be triggered if things are working ok

https://github.com/line/armeria/blob/master/grpc/src/main/java/com/linecorp/armeria/internal/grpc/HttpStreamReader.java#L143

**Comment by tobias-** [comment]
It is not. 

~I'm really not sure about this, but it seems that when doing `HttpServerHandler.handleRequest()` there's a lot of logic used to find the correct Service and call it, but when coming back asynchronously later in `HttpServerHandler.channelInactive()`, that context is gone, although the ChannelHandlerContext is still available.~

~It doesn't seem that the `HttpServerHandler` is shut down properly in the case of closed socket, as `unfinishedRequests` is never counted down to 0 nor is gracefulShutdownSupport.~

All that said (and while true, I believe it to be irrelevant), I looked at how the stream objects/data is propagated, and that uses the `Http2RequestDecoder`. Armeria _does_ get a callback when the client closes the socket as a `Http2RequestDecoder.onStreamRemoved()`, which imho should be propagated a lot more.

Could something as simple as this work? (cloning armeria and testing)
```
     @Override
     public void onStreamRemoved(Http2Stream stream) {
-        requests.remove(stream.id());
+       final DecodedHttpRequest req = requests.remove(stream.id());
+       if (req != null) {
+           req.close();
+       }
     }
```

**Comment by tobias-** [comment]
Ok, that fix works for me. I'm none too fond of signing license agreements, so if somebody who has signed it can open a PR with that, please do. I consider the above code trivial and therefore not copyrightable. Do what you want with it.

I've run all the tests and they pass. This code also makes sure that the `HttpServerHandler` mentioned above is closed _seemingly_ properly, and that propagates into `ResponseObserver.onComplete()` which is exactly where I would expect it to end up, since it isn't an error.

Example commit message:
```
Close the requests associated with killed stream

When a socket is closed, make sure that the outstanding requests
(unfinishedRequests) are closed/terminated as well by closing
the stream when removing it.

Closes #971
```
